spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/TestFunctionCatalog.java

```diff
@@ -0,0 +1,146 @@
+
+  public TestFunctionCatalog() {
+    this.asFunctionCatalog = castToFunctionCatalog(catalogName);
+  }
+
+  @Before
+  public void createDefaultNamespace() {
+    sql("CREATE NAMESPACE IF NOT EXISTS %s", catalogName + ".default");
+  }
+
+  @After
+  public void dropDefaultNamespace() {
+    sql("DROP NAMESPACE IF EXISTS %s", catalogName + ".default");
+  }
+
+  @Test
+  public void testListFunctionsViaCatalog() throws NoSuchNamespaceException {
+    Assertions.assertThat(asFunctionCatalog.listFunctions(EMPTY_NAMESPACE))
+        .anyMatch(func -> "iceberg_version".equalsIgnoreCase(func.name()));
+
+    Assertions.assertThat(asFunctionCatalog.listFunctions(SYSTEM_NAMESPACE))
+        .anyMatch(func -> "iceberg_version".equalsIgnoreCase(func.name()));
+
+    Assert.assertArrayEquals(
+        "Listing functions in an existing namespace that's not system should not throw",
+        new Identifier[0],
+        asFunctionCatalog.listFunctions(DEFAULT_NAMESPACE));
+
+    AssertHelpers.assertThrows(
+        "Listing functions in a namespace that does not exist should throw",
+        NoSuchNamespaceException.class,
+        "Namespace 'db' not found",
+        () -> asFunctionCatalog.listFunctions(DB_NAMESPACE));
+  }
+
+  @Test
+  public void testLoadFunctions() throws NoSuchFunctionException {
+    for (String[] namespace : ImmutableList.of(EMPTY_NAMESPACE, SYSTEM_NAMESPACE)) {
+      Identifier identifier = Identifier.of(namespace, "iceberg_version");
+      UnboundFunction func = asFunctionCatalog.loadFunction(identifier);
+
+      Assertions.assertThat(func)
+          .isNotNull()
+          .isInstanceOf(UnboundFunction.class)
+          .isExactlyInstanceOf(IcebergVersionFunction.class);
+    }
+
+    AssertHelpers.assertThrows(
+        "Cannot load a function if it's not used with the system namespace or the empty namespace",
+        NoSuchFunctionException.class,
+        "Undefined function: default.iceberg_version",
+        () -> asFunctionCatalog.loadFunction(Identifier.of(DEFAULT_NAMESPACE, "iceberg_version")));
+
+    AssertHelpers.assertThrows(
+        "Cannot load a function that does not exist",
+        NoSuchFunctionException.class,
+        "Undefined function: system.undefined_function",
+        () ->
+            asFunctionCatalog.loadFunction(Identifier.of(SYSTEM_NAMESPACE, "undefined_function")));
+
+    AssertHelpers.assertThrows(
+        "Using an undefined function from SQL should fail analysis",
+        AnalysisException.class,
+        "Undefined function",
+        () -> sql("SELECT undefined_function(1, 2)"));
+  }
+
+  @Test
+  public void testIcebergVersionFunctionReturnsCorrectResult() {
+    String buildVersion = IcebergBuild.version();
+
+    // Check that value correctly loads from file, even in codegen etc
+    Assert.assertNotEquals(
+        "iceberg_version should not return undefined",
+        "undefined",
+        scalarSql("SELECT %s.system.iceberg_version()", catalogName));
```

Comment:
Minor: if you have a check for a specific value, that's also a check that it is not some other value. This `assertNotEquals` is redundant.

Reply:
I added this check because the `IcebergBuild.version()` falls back to `undefined` if the file cannot be resolved.

I am essentially testing that the function implementation that gets bound at runtime to `iceberg_version` picks up the correct `IcebergBuild.version()`, though to really test what I want to test it would probably make more sense to add this test in a distributed integration test (i.e. that executors don't receive code that cannot access the value).

Given we already have a test for `IcebergBuild.version()` elsewhere though, this is not necessary.

I'll remove it.

Reply:
Removed!